### PR TITLE
[new release] js_of_ocaml, js_of_ocaml-tyxml, js_of_ocaml-toplevel, js_of_ocaml-ppx_deriving_json, js_of_ocaml-ppx, js_of_ocaml-lwt and js_of_ocaml-compiler (4.0.0)

### PR DESCRIPTION
**Pull request by hhugo**
Compiler from OCaml bytecode to JavaScript

- Project page: <a href="https://ocsigen.org/js_of_ocaml/latest/manual/overview">https://ocsigen.org/js_of_ocaml/latest/manual/overview</a>
- Documentation: <a href="https://ocsigen.org/js_of_ocaml/latest/manual/overview">https://ocsigen.org/js_of_ocaml/latest/manual/overview</a>

## Features/Changes
* Compiler: add --target-env flag, for JS runtime specific compilation targets (ocsigen/js_of_ocaml#1160).
* Compiler: static evaluation of backend_type (ocsigen/js_of_ocaml#1166)
* Compiler: speedup emitting js files (ocsigen/js_of_ocaml#1174)
* Compiler: simplify (a | 0) >>> 0 into (a >>> 0) (ocsigen/js_of_ocaml#1177)
* Compiler: improve static evaluation of cond (ocsigen/js_of_ocaml#1178)
* Compiler: be more consistent dealing with js vs ocaml strings (ocsigen/js_of_ocaml#984)
* Compiler: Compiler: add BigInt to provided symbols (fix ocsigen/js_of_ocaml#1168) (ocsigen/js_of_ocaml#1191)
* Compiler: use globalThis, drop joo_global_object ocsigen/js_of_ocaml#1193
* Compiler: new -Werror flag to turn wanrings into errors (ocsigen/js_of_ocaml#1222)
* Compiler: make the inlining less agressive, reduce size, improve pref (ocsigen/js_of_ocaml#1220)
* Compiler: rename internal library js_of_ocaml-compiler.runtime to js_of_ocaml-compiler.runtime-files
* Lib: new runtime library to improve compatibility with Brr and gen_js_api
* Lib: add messageEvent to Dom_html (ocsigen/js_of_ocaml#1164)
* Lib: add PerformanceObserver API (ocsigen/js_of_ocaml#1164)
* Lib: add CSSStyleDeclaration.{setProperty, getPropertyValue, getPropertyPriority, removeProperty} (ocsigen/js_of_ocaml#1170)
* Lib: make window.{inner,outer}{Width,Height} non-optional
* Lib: introduce Js_of_ocaml.Js_error module, deprecate Js_of_ocaml.Js.Error exception.
* Lib: add deprecation warning for deprecated code
* PPX: json can now be derived for mutable records (ocsigen/js_of_ocaml#1184)
* Runtime: use crypto.getRandomValues when available (ocsigen/js_of_ocaml#1209)
* Misc: move js_of_ocaml-ocamlbuild out to its own repo
* Misc: add support for OCaml 4.14 (ocsigen/js_of_ocaml#1173)

## Bug fixes
* Compiler: fix sourcemap warning for empty cma (ocsigen/js_of_ocaml#1169)
* Compiler: Strengthen bound checks. (ocsigen/js_of_ocaml#1172)
* Compiler: fix `--wrap-with-fun` under node (ocsigen/js_of_ocaml#653, ocsigen/js_of_ocaml#1171)
* Compiler: fix parsing of annotaions in js stubs (ocsigen/js_of_ocaml#1212, fix ocsigen/js_of_ocaml#1213)
* Ppx: allow apostrophe in lident (fix ocsigen/js_of_ocaml#1183) (ocsigen/js_of_ocaml#1192)
* Runtime: fix float parsing in hexadecimal form
* Runtime: fix implementation of caml_js_instanceof
* Graphics: fix mouse_{x,y} (ocsigen/js_of_ocaml#1206)

### CODE REVIEW

File: packages/js_of_ocaml-compiler/js_of_ocaml-compiler.4.0.0/opam [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    jobs
+    "--promote-install-files=false"
+    "@install"
+    "@doc" {with-doc}
+  ]
+  ["dune" "install" "-p" name "--create-install-files" name]
+]
+url {
+  src:
+    "https://github.com/ocsigen/js_of_ocaml/releases/js_of_ocaml-4.0.0.tbz"
```
Comment (kit-ty-kate) [inline packages/js_of_ocaml-compiler/js_of_ocaml-compiler.4.0.0/opam:49]: This archive is not accessible (404). I couldn’t find any releases or tag pointing to a js_of_ocaml 4.0.0 in the repository

Comment (hhugo) [reply packages/js_of_ocaml-compiler/js_of_ocaml-compiler.4.0.0/opam]: fixed

**Comment by hhugo** [comment]
I've fixed the PR

**Review by smorimoto — APPROVED** [review]

**Comment by smorimoto** [comment]
I don't think all the errors are directly caused by this release, so I think we can merge this.

**Comment by kit-ty-kate** [comment]
`js_of_ocaml-webgpu`, `js_of_ocaml-webidl`, `ocsigen-start` and `ocsigen-toolkit` are broken by the new API. They need some upper-bound constraints

**Comment by hhugo** [comment]
I added constraints on ocsigen-toolkit and incr_dom.
ocsigen-start seems to fail because of ocsigen-toolkit
many incr_dom* package seems to fail because of incr_dom

(old version of) js_of_ocaml-webgpu fails because of js_of_ocaml-gpu which I think fails because of menhir + core

**Comment by dbuenzli** [comment]
Can this be merged now ?

**Comment by kit-ty-kate** [comment]
Thanks! Sorry for the delay

**Comment by dbuenzli** [comment]
Merci à tous !

**Comment by dbuenzli** [comment]
Note that the `js_of_ocaml-compiler.4.0.0` packages is still constrained by `"ocaml" {>= "4.04" & < "4.14"}'` is that normal ?

**Comment by kit-ty-kate** [comment]
I opened https://github.com/ocaml/opam-repository/pull/20571 to unlock it

**Pull request merged**